BaseTestHelper.py: Add `BaseTestHelper` class

sangamcse opened the pull request:
`observed_result` is result generated from a bear and `expected_result` is tester's expected result from the bears.

Related to https://gitlab.com/coala/GSoC/GSoC-2018/issues/307
Closes https://github.com/coala/coala/issues/3676

NiklasMM commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,64 @@
+
+class BaseTestHelper:
+    """
+    This is a base class for all tests of coala's testing API.
```
for all tests *Bears*

sangamcse commented:
Updated:

- Docs on [line 6](https://github.com/coala/coala/pull/5496/files#diff-feb1a7b4c2139a13b1c7c4ee1dd84726R6).
- Removed `setUp`, `tearDown`, `setUpClass` and `tearDownClass`.

cc @NiklasMM

kriti21 commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,36 @@
+                        (observed_result, expected_result)
+            if messages:
+                raise AssertionError(messages)
+        else:
```
You can write maybe like 
`if not len(observed_result) == len(expected_result):` This would eliminate one level of indentation atleast. There are so many nested if and for loops.

kriti21 commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,36 @@
+            messages = ''
+            for observed, expected in zip(observed_result, expected_result):
+                if (isinstance(observed, Comparable)
+                        and isinstance(expected, Comparable)) \
```
I might be wrong here but line breaking using `\` is not recommended.

sangamcse replied:
That gives syntax error

IpshitaC commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,41 @@
+    This is a base class for all Bears' tests of coala's testing API.
+    """
+
+    def assertComparableObjectsEqual(self, observed_result, expected_result):
```
Perhaps you could use snake case for the method name, to ensure consistency. `assert_comparable_objects_equal`

IpshitaC replied:
The method name sounds a bit complicated. Is there some way it can be made a bit more intuitive perhaps?

RaiVaibhav replied:
I just gone through the yash Nishar commit of `assertComparableObjectsEqual` and I didn't find any review by John or Makman on name casing, btw for me `assertComparableObjectsEqual` is ok here.

Sorry for nitpicking the comment :stuck_out_tongue:

sangamcse replied:
@RaiVaibhav, it is not necessary to keep same thing in the base class. We can think of better method name here as you can see the previous method name is similar to unittest method name type. But in this project, we are moving towards pytest.

IpshitaC commented on tests/testing/BaseTestHelperTest.py:
```diff
@@ -0,0 +1,31 @@
+
+
+class BaseTestHelperTest(BaseTestHelper):
+    def test_good_assertComparableObjectsEqual(self):
```
A uniform case would be a lot better. `test_good_assert_comparable_objects_equal`

RaiVaibhav replied:
same for me it should be `test_good_assertComparableObjectsEqual` as IMO `assertComparableObjectsEqual` is a method name which together means something

IpshitaC commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,41 @@
+
+    def assertComparableObjectsEqual(self, observed_result, expected_result):
+        """
+        Compares individual fields of the Result object and yields messages in
```
Change to `result objects`

IpshitaC commented on tests/testing/BaseTestHelperTest.py:
```diff
@@ -0,0 +1,31 @@
+    def test_good_assertComparableObjectsEqual(self):
+        self.assertComparableObjectsEqual(['a', 'b'], ['a', 'b'])
+
+    def test_inequality_assertComparableObjectsEqual(self):
```
Same. A uniform snake case for this method name would be a lot better

IpshitaC commented on tests/testing/BaseTestHelperTest.py:
```diff
@@ -0,0 +1,31 @@
+            self.assertComparableObjectsEqual(['a', 'b'], ['a', 'b', None])
+        assert '[\'a\', \'b\'] != [\'a\', \'b\', None]' in str(ex.value)
+
+    def test_not_comparable_assertComparableObjectsEqual(self):
```
Same here. Case of the method name

IpshitaC commented on tests/testing/BaseTestHelperTest.py:
```diff
@@ -0,0 +1,31 @@
+            self.assertComparableObjectsEqual(['a', 'b'], ['a', 'c'])
+        assert '[\'a\', \'b\'] != [\'a\', \'c\']' in str(ex.value)
+
+    def test_comparable_assertComparableObjectsEqual(self):
```
Here too

RaiVaibhav commented:
>Try restarting your Travis build too

Rebuilding will help IMO it's failing on whole coala/coala repo

RaiVaibhav commented:
Ok, :+1:
On Thu, 7 Jun 2018, 11:26 p.m. Sangam Kumar, <notifications@github.com>
wrote:

> *@sangamcse* commented on this pull request.
> ------------------------------
>
> In coalib/testing/BaseTestHelper.py
> <https://github.com/coala/coala/pull/5496#discussion_r193837336>:
>
> > @@ -0,0 +1,41 @@
> +from coala_utils.Comparable import Comparable
> +
> +
> +class BaseTestHelper:
> +    """
> +    This is a base class for all Bears' tests of coala's testing API.
> +    """
> +
> +    def assertComparableObjectsEqual(self, observed_result, expected_result):
>
> @RaiVaibhav <https://github.com/RaiVaibhav>, it is not necessary to keep
> same thing in the base class. We can think of better method name here as
> you can see the previous method name is similar to unittest method name
> type. But in this project, we are moving towards pytest.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coala/coala/pull/5496#discussion_r193837336>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AVPxJlchCvT1VEfLbw_Q206SnA2I108vks5t6WlmgaJpZM4Ua7p3>
> .
>

IpshitaC commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,44 @@
+                    except AssertionError as ex:
+                        messages += (str(ex) + '\n\n')
+            else:
+                assert observed_result == expected_result, '%s != %s' % (
```
I think this is good enough.

Makman2 replied:
so no unittest style asserts any more?

sangamcse replied:
Yes 😅

sangamcse commented:
@gitmate-bot rebase

jayvdb commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,44 @@
+                                  expected_result):
+        """
+        Compares individual fields of the result objects and yields messages
+        in case of an attribute mismatch.
```
'attribute mismatch' is unclear.

Is it attribute *type* mismatch, or attribute *value* mismatch?

When I read `assert_comparable_objects`, I assume it means it checks that the *types* are comparable, but does not check that they are the same.

If it is about the values, then the name should include they key word `equals`, like [`assertCountEqual`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual)

However, by reading the code I can see that `comparable` here refers to `coala_utils.Comparable`.  That needs to be described in the docstring.

fwiw, most of this logic should be put into `Comparable.__eq__`, and similar methods.

Then the code here would simple use `!=` , and raise asserts accordingly.

jayvdb commented on coalib/testing/BaseTestHelper.py:
```diff
@@ -0,0 +1,56 @@
+    This is a base class for all Bears' tests of coala's testing API.
+    """
+
+    def _assert_comparable_equal(self,
```
module level scope means move it outside the class.

sangamcse replied:
Ohk. Got it.

jayvdb commented:
ack 936a5d8

jayvdb commented:
@gitmate-bot ff
